pkg/webhook/stats_reporter.go

```diff
@@ -0,0 +1,127 @@
+package webhook
+
+import (
+	"context"
+	"strconv"
+	"time"
+
+	"github.com/open-policy-agent/gatekeeper/pkg/metrics"
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+)
+
+const (
+	requestCountName     = "request_count"
+	requestLatenciesName = "request_latencies"
+)
+
+var (
+	requestCountM = stats.Int64(
+		requestCountName,
+		"The number of requests that are routed to webhook",
+		stats.UnitDimensionless)
+	responseTimeInMsecM = stats.Float64(
+		requestLatenciesName,
+		"The response time in milliseconds",
+		stats.UnitMilliseconds)
+
+	requestOperationKey  = tag.MustNewKey("request_operation")
+	kindGroupKey         = tag.MustNewKey("kind_group")
+	kindVersionKey       = tag.MustNewKey("kind_version")
+	kindKindKey          = tag.MustNewKey("kind_kind")
+	resourceGroupKey     = tag.MustNewKey("resource_group")
+	resourceVersionKey   = tag.MustNewKey("resource_version")
+	resourceResourceKey  = tag.MustNewKey("resource_resource")
+	resourceNameKey      = tag.MustNewKey("resource_name")
+	resourceNamespaceKey = tag.MustNewKey("resource_namespace")
+	admissionAllowedKey  = tag.MustNewKey("admission_allowed")
+)
+
+func init() {
+	register()
+}
+
+// StatsReporter reports webhook metrics
+type StatsReporter interface {
+	ReportRequest(request *admissionv1beta1.AdmissionRequest, response *admissionv1beta1.AdmissionResponse, d time.Duration) error
+}
+
+// reporter implements StatsReporter interface
+type reporter struct {
+	ctx context.Context
+}
+
+// NewStatsReporter creaters a reporter for webhook metrics
+func NewStatsReporter() (StatsReporter, error) {
+	ctx, err := tag.New(
+		context.Background(),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &reporter{ctx: ctx}, nil
+}
+
+// Captures req count metric, recording the count and the duration
+func (r *reporter) ReportRequest(req *admissionv1beta1.AdmissionRequest, resp *admissionv1beta1.AdmissionResponse, d time.Duration) error {
+	ctx, err := tag.New(
+		r.ctx,
+		tag.Insert(requestOperationKey, string(req.Operation)),
```

Comment:
For the same scalability/security/privacy reasons as above, we should not tag the metric with sensitive or potentially unbounded information.

Reply:
should we leave any of the tags or remove all of them?

Reply:
I don't think tagging the GVK of the admission request would end up being useful. The resource name, namespace, and GVK are unbounded and private info. The result of the admission is definitely useful, but see my comment above.

Reply:
removed request operations